File: Java/Tests/jdi-uitest-wintests/src/test/java/com/epam/jdi/uitests/testing/unittests/tests/complex/ChecklistTests.java
package com.epam.jdi.uitests.testing.unittests.tests.complex;

import com.epam.jdi.uitests.core.interfaces.complex.ICheckList;
import com.epam.jdi.uitests.testing.unittests.InitTests;
import com.epam.jdi.uitests.testing.unittests.enums.Nature;
import org.openqa.selenium.By;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import static com.epam.commons.LinqUtils.first;
import static com.epam.jdi.uitests.core.preconditions.PreconditionsState.isInState;
import static com.epam.jdi.uitests.testing.unittests.enums.Preconditions.METALS_AND_COLORS_PAGE;
import static com.epam.jdi.uitests.testing.unittests.pageobjects.EpamJDIApplication.actionsLog;
import static com.epam.jdi.uitests.testing.unittests.pageobjects.EpamJDIApplication.metalsColorsPage;
import static com.epam.jdi.uitests.testing.unittests.tests.complex.CommonActionsData.checkAction;
import static com.epam.jdi.uitests.win.settings.WinSettings.getDriver;
import static com.epam.web.matcher.testng.Assert.*;
import static java.util.Arrays.asList;

/**
 * Created by Roman_Iovlev on 9/15/2015.
 */
public class ChecklistTests extends InitTests {
    private static final List<String> natureOptions = asList("Water", "Earth", "Wind", "Fire");
    private static final String allValues = "Water, Earth, Wind, Fire";

    private ICheckList<Nature> nature() {
        return metalsColorsPage.nature;
    }

    @BeforeMethod
    public void before(Method method) throws IOException {
        isInState(METALS_AND_COLORS_PAGE);
    }

    private void checkAllChecked() {
        assertTrue(first(getDriver().findElements(By.cssSelector("#elements-checklist input")),
                el -> el.getAttribute("checked") != null) != null);
    }

    @Test
    public void selectStringTest() {
        nature().select("Fire");
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void selectIndexTest() {
        nature().select(4);
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void selectEnumTest() {
        nature().select(Nature.FIRE);
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void select2StringTest() {
        nature().select("Water", "Fire");
        checkAction("Fire: condition changed to true");
        assertContains(() -> (String) actionsLog.getTextList().get(1), "Water: condition changed to true");

    }
private List<String> ls() { return new ArrayList<>(); }

    @Test
    public void select2IndexTest() {
        nature().select(1, 4);
        checkAction("Fire: condition changed to true");
        assertContains(() -> actionsLog.getTextList().get(1), "Water: condition changed to true");
    }

    @Test
    public void select2EnumTest() {
        nature().select(Nature.WATER, Nature.FIRE);
        checkAction("Fire: condition changed to true");
        assertContains(() -> actionsLog.getTextList().get(1), "Water: condition changed to true");
    }

    @Test
    public void checkStringTest() {
        nature().check("Fire");
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void checkIndexTest() {
        nature().check(4);
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void checkEnumTest() {
        nature().check(Nature.FIRE);
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void check2StringTest() {
        nature().check("Water", "Fire");
        checkAction("Fire: condition changed to true");
        assertContains(() -> actionsLog.getTextList().get(1), "Water: condition changed to true");

    }

    @Test
    public void check2IndexTest() {
        nature().check(1, 4);
        checkAction("Fire: condition changed to true");
        assertContains(() -> actionsLog.getTextList().get(1), "Water: condition changed to true");
    }

    @Test
    public void check2EnumTest() {
        nature().check(Nature.WATER, Nature.FIRE);
        checkAction("Fire: condition changed to true");
        assertContains(() -> actionsLog.getTextList().get(1), "Water: condition changed to true");
    }

    @Test
    public void selectAllTest() {
        nature().selectAll();
        List<String> log = actionsLog.getTextList();
        assertContains(log.get(3), "Water: condition changed to true");
        assertContains(log.get(2), "Earth: condition changed to true");
        assertContains(log.get(1), "Wind: condition changed to true");
        assertContains(log.get(0), "Fire: condition changed to true");
        checkAllChecked();
    }

    @Test
    public void checkAllTest() {
        nature().checkAll();
        List<String> log = actionsLog.getTextList();
        assertContains(log.get(3), "Water: condition changed to true");
        assertContains(log.get(2), "Earth: condition changed to true");
        assertContains(log.get(1), "Wind: condition changed to true");
        assertContains(log.get(0), "Fire: condition changed to true");
        checkAllChecked();
    }

    @Test
    public void clearAllTest() {
        nature().checkAll();
        checkAllChecked();
        nature().clear();
        checkAllChecked(); // isDisplayed not defined
    }

    @Test
    public void uncheckAllTest() {
        nature().checkAll();
        checkAllChecked();
        nature().uncheckAll();
        checkAllChecked(); // isDisplayed not defined
    }

    @Test
    public void getOptionsTest() {
        listEquals(nature().getOptions(), natureOptions);
    }

    @Test
    public void getNamesTest() {
        listEquals(nature().getNames(), natureOptions);
    }

    @Test
    public void getValuesTest() {
        listEquals(nature().getValues(), natureOptions);
    }

    @Test
    public void getOptionsAsTextTest() {
        areEquals(nature().getOptionsAsText(), allValues);
    }

    @Test
    public void setValueTest() {
        nature().setValue("Fire");
        checkAction("Fire: condition changed to true");
    }

    @Test
    public void getNameTest() {
        areEquals(nature().getName(), "Nature");
    }

    @Test
    public void areSelectedTest() {
        listEquals(nature().areSelected(), new ArrayList<>());// isDisplayed not defined
    }

    @Test
    public void areDeselectedTest() {
        listEquals(nature().areDeselected(), natureOptions);// isDisplayed not defined
    }

    @Test
    public void getValueTest() {
        areEquals(nature().getValue(), "");// isDisplayed not defined
    }
}

File: C#.Net/Tests/JDI UIWebTests/JDI UIWebTests/Tests/Complex/CheckListTests.cs
using Epam.JDI.Core.Interfaces.Complex;
using NUnit.Framework;
using System.Collections.Generic;
using static Epam.JDI.Core.Settings.JDISettings;
using static JDI_UIWebTests.UIObjects.TestSite;
using JDI_UIWebTests.Enums;
using static JDI_UIWebTests.Tests.Complex.CommonActionsData;
using JDI_Matchers.NUnit;
using OpenQA.Selenium;
using System.Linq;


namespace JDI_UIWebTests.Tests.Complex
{
    public class CheckListTests
    {
        private static readonly IList<string> NATURE_OPTIONS = new List<string>{"Water", "Earth", "Wind", "Fire" };           
        private static readonly string ALL_VALUES = "Water, Earth, Wind, Fire";

        private ICheckList<Elements> _nature()
        {
            return MetalsColorsPage.Elements;
        }

        private void _checkAllIsChecked(bool isChecked)
        {
            IList<bool> checkedElems = HomePage.WebDriver.
                FindElements(By.CssSelector("#elements-checklist input")).
                Select(e => e.GetAttribute("checked") != null).ToList();
            new Check("Check that all checkbox elements checked = " + isChecked).
                IsTrue(checkedElems.All(e => e == isChecked));            
        }

        private void _clearCheckBoxBlock() {
            IList<IWebElement> inputElems = HomePage.WebDriver.FindElements(By.CssSelector("#elements-checklist input"));
            IList<IWebElement> labelsElems = HomePage.WebDriver.FindElements(By.CssSelector(".checkbox>label"));

            for (int i = 0; i < inputElems.Count; i++)
            {
                if ((inputElems[i].GetAttribute("checked") != null) && (inputElems[i].GetAttribute("checked") == "true"))
                {
                    labelsElems[i].Click();
                }
            }
        }

        private void _cheсkAllLogMessages(IList<string> logLines) {
            //TO_DO: replace with TextList.Texts when will be fixed
            //var texts = ActionsLog.Texts; 
            IList<string> log = HomePage.WebDriver.FindElements(By.CssSelector(".logs li")).Select(e => e.Text).ToList();            
            for (int i = 0; i < log.Count; i++)
            {
                JDI_Matchers.NUnit.Assert.Contains(log[i], logLines[i] + ": condition changed to true");
            }            
        }

        [SetUp]
        public void SetUp()
        {
            Logger.Info("Navigating to Metals and Colors page.");
            MetalsColorsPage.Open();
            MetalsColorsPage.CheckTitle();
            MetalsColorsPage.IsOpened();
            Logger.Info("Setup method finished");
            Logger.Info("Start test: " + TestContext.CurrentContext.Test.Name);
        }

        [Test]
        public void SelectStringTest()
        {
            _nature().Select("Fire");
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void SelectIndexTest()
        {
            _nature().Select(4);
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void SelectEnumTest()
        {
            _nature().Select(Elements.Fire);
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void Select2StringTest()
        {
            _nature().Select("Water", "Fire");
            CheckAction("Fire: condition changed to true");            
            string asd = ActionsLog.TextElements.First().Value;
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void Select2IndexTest()
        {
            _nature().Select(1, 4);
            CheckAction("Fire: condition changed to true");
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void Select2EnumTest()
        {
            _nature().Select(Elements.Water, Elements.Fire);
            CheckAction("Fire: condition changed to true");
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void CheckStringTest()
        {
            _nature().Check("Fire");
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void CheckIndexTest()
        {
            _nature().Check(4);
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void CheckEnumTest()
        {
            _nature().Check(Elements.Fire);
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void Check2StringTest()
        {
            _nature().Check("Water", "Fire");
            CheckAction("Fire: condition changed to true");
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void Check2IndexTest()
        {
            _nature().Check(1, 4);
            CheckAction("Fire: condition changed to true");
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void Check2EnumTest()
        {
            _nature().Check(Elements.Water, Elements.Fire);
            CheckAction("Fire: condition changed to true");
            _cheсkAllLogMessages(new List<string>() { "Fire", "Water" });
        }

        [Test]
        public void SelectAllTest()
        {
            _nature().SelectAll();            
            _cheсkAllLogMessages(new List<string>() { "Fire", "Wind", "Earth", "Water" });
            _checkAllIsChecked(true);         

        }

        [Test]
        public void CheckAllTest()
        {
            _nature().CheckAll();
            _cheсkAllLogMessages(new List<string>() { "Fire", "Wind", "Earth", "Water" });
            _checkAllIsChecked(true);
        }

        [Test]
        public void ClearAllTest()
        {
            _nature().CheckAll();
            _checkAllIsChecked(true);            
            _clearCheckBoxBlock();            
            _checkAllIsChecked(false);  
        }

        [Test]
        public void GetOptionsTest()
        {
            new Check().CollectionEquals(_nature().Options, NATURE_OPTIONS);            
        }

        [Test]
        public void GetNamesTest()
        {
            new Check().CollectionEquals(_nature().Names, NATURE_OPTIONS);
        }

        [Test]
        public void GetValuesTest()
        {
            new Check().CollectionEquals(_nature().Values, NATURE_OPTIONS);
        }

        [Test]
        public void GetOptionsAsTextTest()
        {
            new Check().AreEquals(_nature().OptionsAsText.ToString(), ALL_VALUES);            
        }

        [Test]
        public void SetValueTest()
        {
            _nature().Value = "Fire";                
            CheckAction("Fire: condition changed to true");
        }

        [Test]
        public void GetNameTest()
        {            
            new Check().AreEquals(_nature().Name, "Elements");            
        }

        // TODO: fix incorrect work of AreSelected method. It is always return empty collection
        /*
        [Test]
        public void AreSelectedTest()
        {
            new Check().CollectionEquals(_nature().AreSelected(), new List<string>() );
        }


        //TO_DO fix incorrect work of AreSelected method.It is always return empty collection
        [Test]
        public void AreDeselectedTest()
        {
            new Check().CollectionEquals(_nature().AreDeselected(), NATURE_OPTIONS);
        }
        */

        [Test]
        public void GetValueTest()
        {
            new Check().AreEquals(_nature().Value, "");                  
        }
    }
}
